components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Storage/PublishedNodesJobConverter.cs

```diff
@@ -71,130 +71,177 @@ public class PublishedNodesJobConverter {
+                // note: do not remove 'unnecessary' .ToList(),
+                // the grouping of operations improves perf by 30%
+                Stopwatch sw = new Stopwatch();
+                sw.Start();
+                // Group by connection
+                var group = items.GroupBy(
+                    item => new ConnectionModel {
                         OperationTimeout = legacyCliModel.OperationTimeout,
                         Id = item.DataSetWriterId,
                         Group = item.DataSetWriterGroup,
                         Endpoint = new EndpointModel {
                             Url = item.EndpointUrl.OriginalString,
                             SecurityMode = item.UseSecurity == false &&
-                                item.OpcAuthenticationMode != OpcAuthenticationMode.UsernamePassword ?
-                                    SecurityMode.None : SecurityMode.Best
+                            item.OpcAuthenticationMode != OpcAuthenticationMode.UsernamePassword ?
+                            SecurityMode.None : SecurityMode.Best
                         },
                         User = item.OpcAuthenticationMode != OpcAuthenticationMode.UsernamePassword ?
-                                null : ToUserNamePasswordCredentialAsync(item).Result,
-                        
+                        null : ToUserNamePasswordCredentialAsync(item).Result,
+
                     },
-                        // Select and batch nodes into published data set sources
-                        item => GetNodeModels(item, legacyCliModel.ScaleTestCount.GetValueOrDefault(1)),
-                        // Comparer for connection information
-                        new FuncCompare<ConnectionModel>((x, y) => x.IsSameAs(y)))
-                    .Select(group => group
-                        // Flatten all nodes for the same connection and group by publishing interval
-                        // then batch in chunks for max 1000 nodes and create data sets from those.
-                        .Flatten()
-                        .GroupBy(n => n.OpcPublishingInterval)
-                        .SelectMany(n => n
-                            .Distinct((a, b) => a.Id == b.Id && a.DisplayName == b.DisplayName && a.DataSetFieldId == b.DataSetFieldId &&
-                                        a.OpcSamplingInterval == b.OpcSamplingInterval)
-                            .Batch(1000))
-                        .Select(opcNodes => new PublishedDataSetSourceModel {
+                    // Select and batch nodes into published data set sources
+                    item => GetNodeModels(item, legacyCliModel.ScaleTestCount.GetValueOrDefault(1)),
+                    // Comparer for connection information
+                    new FuncCompare<ConnectionModel>((x, y) => x.IsSameAs(y))
+                ).ToList();
+                _logger.Information("Created {n} groups in {elapsed}", group.Count, sw.Elapsed);
```

Comment:
>i [](http://example.com/codeflow?start=32&length=1)

This log isn't so relevant for the customer, please change to Debug log level

Reply:
removed